definitions/output/all/backfill_requests.js

```diff
@@ -0,0 +1,209 @@
+    return ""; // Return an empty string in case of any errors
+  }
+""";
+
+CREATE TEMP FUNCTION get_type(mime_typ STRING, ext STRING)
+RETURNS STRING
+LANGUAGE js
+AS """
+  try {
+    mime_typ = mime_typ.toLowerCase();
+
+    // Order by most unique types first
+    const uniqueTypes = ["font", "css", "image", "script", "video", "audio", "xml"];
+    for (let typ of uniqueTypes) {
+      if (mime_typ.includes(typ)) {
+        return typ;
+      }
+    }
+
+    // Special cases
+    if (mime_typ.includes("json") || ["js", "json"].includes(ext)) {
+      return "script";
+    } else if (["eot", "ttf", "woff", "woff2", "otf"].includes(ext)) {
+      return "font";
+    } else if (
+      ["png", "gif", "jpg", "jpeg", "webp", "ico", "svg", "avif", "jxl", "heic", "heif"].includes(ext)
+    ) {
+      return "image";
+    } else if (ext === "css") {
+      return "css";
+    } else if (ext === "xml") {
+      return "xml";
+    } else if (
+      ["mp4", "webm", "ts", "m4v", "m4s", "mov", "ogv", "swf", "f4v", "flv"].includes(ext) ||
+      ["flash", "webm", "mp4", "flv"].some(typ => mime_typ.includes(typ))
+    ) {
+      return "video";
+    } else if (mime_typ.includes("wasm") || ext === "wasm") {
+      return "wasm";
+    } else if (mime_typ.includes("html") || ["html", "htm"].includes(ext)) {
+      return "html";
+    } else if (mime_typ.includes("text")) {
+      // Put "text" last because it is often misused, so extension should take precedence.
+      return "text";
+    } else {
+      return "other";
+    }
+  } catch (e) {
+    return "other"; // Return "other" if there's any error
+  }
+""";
+
+CREATE TEMP FUNCTION parse_headers(headers JSON)
+RETURNS ARRAY<STRUCT<name STRING, value STRING>>
+LANGUAGE js
+AS """
+  try {
+    return headers.map(header => {
+      return { name: header.name, value: header.value };
+    });
+  } catch (e) {
+    return [];
+  }
+""";
+
+INSERT INTO all_dev.requests_stable
+SELECT
+  DATE('${iteration.date}') AS date,
+  '${iteration.client}' AS client,
+  requests.page AS page,
+  TRUE AS is_root_page,
+  requests.page AS root_page,
+  crux.rank AS rank,
+  requests.url AS url,
+  IF(
+    SAFE_CAST(payload._request_type AS STRING) = "Document" AND
+      MIN(SAFE_CAST(payload._index AS INT64)) OVER (PARTITION BY requests.page) = SAFE_CAST(payload._index AS INT64),
+    TRUE,
+    FALSE
+  ) AS is_main_document,
+  get_type(SAFE_CAST(payload.response.content.mimeType AS STRING), ext_from_url) AS type,
+  SAFE_CAST(payload._index AS INT64) AS index,
+  JSON_REMOVE(
+    payload,
+    '$._headers',
+    '$.request.headers',
+    '$.response.headers'
+  ) AS payload,
+  TO_JSON( STRUCT(
+    payload.time,
+    payload._method AS method,
+    NULL AS redirectUrl,
+    IFNULL(SAFE_CAST(payload._protocol AS STRING), SAFE_CAST(payload.request.httpVersion AS STRING) AS reqHttpVersion,
+    payload.request.headersSize AS reqHeadersSize,
+    payload.request.bodySize AS reqBodySize,
+    NULL AS reqCookieLen,
```

Comment:
I couldn't find a way to join `summary_requests` table, so some summary metrics are empty.

There needs to be unique page and request identifier. `url` is not enough:
```
LEFT JOIN (
  SELECT
    url,
    redirectUrl,
    reqCookieLen,
    respCookieLen,
    expAge,
    format
  FROM summary_requests.${constants.fnDateUnderscored(iteration.date)}_${iteration.client} ${constants.devTABLESAMPLE}
) AS summary_requests
ON requests.url = summary_requests.url
```

Reply:
@pmeenan What is a unique key to join summary_requests and requests before 2022?

Reply:
I don't know that those were ever joined with each other. When joined with the page tables, the `page` (requests) or `pageid` (summary_requests) were joined with the `pageid` or `url` of the respective pages tables.